src/utilities/api.rest.ts

```diff
@@ -0,0 +1,109 @@
+import axios, {
+  AxiosError,
+  AxiosResponse,
+  Method,
+} from 'axios';
+import {getToken} from "../helper/auth";
+
+class Api {
+  private base_url = 'https://api.zid.sa/v1' as string;
+  private headers = {} as { [key: string]: string };
+  private route = '' as string;
+  private method = 'GET' as Method;
+  private body: unknown;
+  private token = '';
+  private params = '' as string;
+  private key = '' as string;
+
+  constructor() {
+    this.token = getToken();
+  }
+
+
+  public addParams(params?: { key: string; value: string | string[] }[]) {
+    params &&
+      params.length > 0 &&
+      params?.map(
+        ({ key, value }) =>
+          (this.params += `${this.params === '' ? '?' : '&'}${key}=${
+            Array.isArray(value) ? value.join(',') : value
+          }`)
+      );
+    return this;
+  }
+
+  public addKey(key: string) {
+    this.key = this.key ? `${this.key}/${key}` : `/${key}`;
+    return this;
+  }
+
+  public addHeaders(headers: { key: string; value: string }[]) {
+    headers.map(({ key, value }) => (this.headers[key] = value));
+    return this;
+  }
+
+  public addUserToken() {
+    this.headers['x-partner-token'] = `${this.token}`;
+    return this;
+  }
+
+  public addRoute(endpoint: string) {
+    this.route = `${this.base_url}${endpoint}`;
+    return this;
+  }
+
+  public addBody(body: unknown) {
+    this.body = body;
+    return this;
+  }
+
+  public post() {
+    this.method = 'POST';
+    return this;
+  }
+
+  public get() {
+    this.method = 'GET';
+    return this;
+  }
+
+  public put() {
+    this.method = 'PUT';
+    return this;
+  }
+
+  public delete() {
+    this.method = 'DELETE';
+    return this;
+  }
+
+  public send(): Promise<unknown> {
+    const url = `${this.route}${this.key}${this.params}`;
```

Comment:
I think it is better to use axios methods like `axios.get` to make requests to get the advantage of passing query params in a better way like this:
`axios.get('/api', {
  params: {
    id: 123,
    name: 'John Doe'
  }
})`

Reply:
but as you can see, this is a dynamic function, that can be used in all the requests types using a smart way.
where you can pass the parameters using `addParams()` function.

Reply:
OK, you have a good point. 👍